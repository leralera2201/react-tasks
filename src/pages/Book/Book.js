import React, { Component } from "react";
import {Button, Card, CardBody, CardSubtitle, CardText, CardTitle} from "reactstrap";
import { connect } from "react-redux";
import moment from "moment";

import { Loader, Error } from "../../components";

import {deleteBookStart, fetchBookStart} from "./actions/book.actions";
import {
  bookDeleteIsStatusSuccess,
  bookFetchDataSelector,
  bookFetchErrorSelector,
  bookFetchIsStatusInProgress
} from "./selectors/book.selectors";


class Book extends Component {
  componentDidMount() {
    const {
      getBookById,
      match: {
        params: { id },
      },
    } = this.props;
    getBookById(id);
  }


  deleteBook = id => {
    const { deleteBook, history } = this.props;
    const confirmation = window.confirm('Are you sure?');
    if(confirmation){
      deleteBook(id);
      history.push({
        pathname: '/books',
        state: { withoutFetch: true }
      });
    }
  }

  updateBook = id => {
    const { book, history } = this.props;
    history.push({
      pathname: `/books/${id}/update`,
      state: { book }
    });
  }

  render() {
    const { loading, book, error } = this.props;

    return (
      <div className="container pt-5">
        {loading && <Loader />}
        {!loading && !error && book && (
          <Card>
            <CardBody>
              <CardTitle tag="h3">{book.title}</CardTitle>
              <CardSubtitle tag="h6" className="mb-2 text-muted">
                {book.description}
              </CardSubtitle>
              <CardText>{book.excerpt}</CardText>
              <CardText tag="h5">Page count: {book.pageCount}</CardText>
              <CardText>
                <small className="text-muted">
                  {moment(book.publishDate).fromNow()}
                </small>
              </CardText>
              <Button onClick={() => this.deleteBook(book.id)}>
                Delete this book
              </Button>
              <Button onClick={() => this.updateBook(book.id)}>
                Update this book
              </Button>
            </CardBody>
          </Card>
        )}
        {!error && !loading && !book && <div>No book</div>}
        {error && <Error error={error} />}
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  book: bookFetchDataSelector(state),
  loading: bookFetchIsStatusInProgress(state),
  error: bookFetchErrorSelector(state),
  deleteSuccess: bookDeleteIsStatusSuccess(state)
});

const mapDispatchToProps = (dispatch) => ({
  getBookById: (id) => dispatch(fetchBookStart(id)),
  deleteBook: (id) => dispatch(deleteBookStart(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(Book);
