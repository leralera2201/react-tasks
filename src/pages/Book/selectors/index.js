import { createSelector } from "reselect";

const bookSelector = (state) => state.book;

export const bookItemSelector = createSelector(
  bookSelector,
  ({ data }) => data
);

export const bookLoadingSelector = createSelector(
  bookSelector,
  ({ loading }) => loading
);

export const bookErrorSelector = createSelector(
  bookSelector,
  ({ error }) => error
);
