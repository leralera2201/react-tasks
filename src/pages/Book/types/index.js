export const BOOK_REQUEST = "BOOK_REQUEST";
export const SET_BOOK = "SET_BOOK";
export const BOOK_ERROR = "BOOK_ERROR";
