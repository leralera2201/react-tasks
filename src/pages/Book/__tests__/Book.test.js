import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

import { Book } from "../Book";

describe("book", () => {
  const props = {
    loading: true,
    error: null,
    book: null,
    deleteSuccess: false,
    getBookById: () => {},
    deleteBook: () => {},
    match: {
      params: {},
    },
    history: { push: jest.fn(), location: {}, listen: jest.fn() },
  };

  it("render with loading", () => {
    const wrapper = shallow(<Book {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("render no book", () => {
    props.loading = false;
    const wrapper = shallow(<Book {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("render book", () => {
    props.book = {
      title: "d",
      description: "d",
      excerpt: "d",
      publishDate: "2000-20-20",
      pageCount: 2,
      id: "1",
    };
    const wrapper = shallow(<Book {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("update book", () => {
    const wrapper = shallow(<Book {...props} />);
    const button = wrapper.find("#update-book").first();
    button.simulate("click");
    expect(props.history.push.mock.calls[0][0].pathname).toEqual(
      `/books/${props.book.id}/update`
    );
  });

  it("delete book", () => {
    window.confirm = jest.fn(() => true);
    const wrapper = shallow(<Book {...props} />);
    const button = wrapper.find("#delete-book").first();
    button.simulate("click");
    expect(window.confirm).toBeCalled();
    expect(props.history.push.mock.calls[0][0].pathname).toEqual(`/books`);
  });

  it("render with error", () => {
    props.error = "Some error";
    const wrapper = shallow(<Book {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

// describe('testing Book component', () => {
//   const props = {
//     loading: true,
//     error: null,
//     book: null,
//     deleteSuccess: false,
//     getBookById: () => {},
//     deleteBook: () => {},
//     match: {
//       params: {},
//     },
//     history: { push: jest.fn(), location: {}, listen: jest.fn() },
//   };
//
//   let wrapper;
//
//   beforeEach(() => {
//     wrapper = shallow(<Book {...props} />);
//   });
//
//   it('renders correctly', () => {
//     expect(wrapper).toMatchSnapshot();
//   });
//
//   it('render with loading', () => {
//     expect(wrapper.find(Loader)).toHaveLength(1);
//   });
//
//   it('render no book', () => {
//     wrapper.setProps({ loading: false });
//     expect(wrapper.find('#no-book')).toHaveLength(1);
//   });
//
//   // it('update book', () => {
//   //   const button = wrapper.find('#update-book');
//   //   button.simulate('click');
//   //   expect(props.history.push.mock.calls[0][0].pathname).toEqual(
//   //     `/books/${props.book.id}/update`
//   //   );
//   // });
//
//   // it('delete book', () => {
//   //   window.confirm = jest.fn(() => true);
//   //   const button = wrapper.find('#delete-book');
//   //   button.simulate('click');
//   //   expect(window.confirm).toBeCalled();
//   //   expect(props.history.push.mock.calls[0][0].pathname).toEqual(`/books`);
//   // });
//
//   it('render with error', () => {
//     wrapper.setProps({ error: 'Some error' });
//     expect(wrapper.find(Error)).toHaveLength(1);
//   });
// });
