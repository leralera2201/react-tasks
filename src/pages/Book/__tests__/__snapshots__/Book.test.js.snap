// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`book render book 1`] = `
<div
  className='container pt-5'
>
  <Card
    tag='div'
  >
    <CardBody
      tag='div'
    >
      <CardTitle
        tag='h3'
      >
        d
      </CardTitle>
      <CardSubtitle
        className='mb-2 text-muted'
        tag='h6'
      >
        d
      </CardSubtitle>
      <CardText
        tag='p'
      >
        d
      </CardText>
      <CardText
        tag='h5'
      >
        Page count: 
        2
      </CardText>
      <CardText
        tag='p'
      >
        <small
          className='text-muted'
        >
          Invalid date
        </small>
      </CardText>
      <Button
        color='secondary'
        id='delete-book'
        onClick={[Function]}
        tag='button'
      >
        Delete this book
      </Button>
      <Button
        color='secondary'
        id='update-book'
        onClick={[Function]}
        tag='button'
      >
        Update this book
      </Button>
    </CardBody>
  </Card>
</div>
`;

exports[`book render no book 1`] = `
<div
  className='container pt-5'
>
  <div
    id='no-book'
  >
    No book
  </div>
</div>
`;

exports[`book render with error 1`] = `
<div
  className='container pt-5'
>
  <Error
    error='Some error'
  />
</div>
`;

exports[`book render with loading 1`] = `
<div
  className='container pt-5'
>
  <Loader />
</div>
`;
